core/src/main/scala/me/wojnowski/oidc4s/IdTokenVerifier.scala

```diff
@@ -104,49 +103,74 @@ object IdTokenVerifier {
+          header     <- EitherT.fromEither(decodeHeader(headerJson))
+          publicKey  <- EitherT(publicKeyProvider.getKey(header.keyId).map(_.leftMap(IdTokenVerifier.Error.CouldNotFindPublicKey.apply)))
           result     <- EitherT.fromEither {
-                          decodeAndVerifyToken[(A, IdTokenClaims)](rawToken, javaClock, publicKey)
-                            .flatMap { case (claims, standardClaims) =>
-                              ensureExpectedIssuer(tokenIssuer = standardClaims.issuer, expectedIssuer = issuer)
-                                .leftWiden[IdTokenVerifier.Error]
-                                .flatTap { _ =>
-                                  standardClaimsCheck(standardClaims)
-                                }
-                                .as(claims)
-                            }
+                          decodeJwtAndVerifySignature[A](rawToken, publicKey, header).flatMap { case (claims, standardClaims) =>
+                            List(
+                              ensureNotExpired(now, standardClaims.expiration),
+                              ensureExpectedIssuer(tokenIssuer = standardClaims.issuer, expectedIssuer = issuer),
+                              standardClaimsCheck(standardClaims)
+                            ).sequence.as(claims)
+                          }
                         }
         } yield result
       }.value
 
-      private def decodeAndVerifyToken[A: JsonDecoder](
-        rawToken: String,
-        javaClock: JavaClock,
-        publicKey: PublicKey
-      ): Either[Error, A] =
-        Jwt(javaClock)
-          .decodeRaw(rawToken, publicKey, supportedAlgorithms)
-          .toEither
-          .leftMap[Error](throwable => JwtVerificationError(throwable))
-          .flatMap { rawClaims =>
-            JsonDecoder[A]
-              .decode(rawClaims)
-              .leftMap(IdTokenVerifier.Error.CouldNotDecodeClaim.apply)
-          }
-
       private def ensureExpectedIssuer(tokenIssuer: Issuer, expectedIssuer: Issuer): Either[Error.UnexpectedIssuer, Unit] =
         Either.cond(expectedIssuer === tokenIssuer, (), IdTokenVerifier.Error.UnexpectedIssuer(tokenIssuer, expectedIssuer))
 
-      private def extractKid(headerJson: String): Either[CouldNotExtractKeyId.type, String] =
+      private def ensureNotExpired(now: Instant, expiresAt: Instant): Either[Error.TokenExpired, Unit] =
+        Either.raiseWhen(expiresAt.isBefore(now))(TokenExpired(since = expiresAt))
+
+      private def decodeHeader(headerJson: String): Either[CouldNotDecodeHeader, JwtHeader] =
         JsonDecoder[JwtHeader]
           .decode(headerJson)
-          .toOption
-          .flatMap(_.keyId)
-          .toRight(CouldNotExtractKeyId)
+          .leftMap(CouldNotDecodeHeader.apply)
+
+      private def decodeJwtAndVerifySignature[A: ClaimsDecoder](rawToken: String, key: PublicKey, header: JwtHeader)
```

Comment:
Looks good overall. I'd just suggest adding a reference to docs (https://openid.net/specs/openid-connect-core-1_0.html#ImplicitIDTValidation) in comment, useful when refactoring :wink: